chain/chunks/src/test/basic.rs

```diff
@@ -68,9 +117,142 @@ fn test_basic() {
     }
     match &test.data.client_events[1] {
         ShardsManagerResponse::ChunkCompleted { partial_chunk, shard_chunk } => {
-            assert_eq!(partial_chunk.parts().len(), fixture.all_part_ords.len());
+            assert_eq!(partial_chunk.parts().len(), chunk.part_ords().len());
             assert!(shard_chunk.is_some());
         }
         _ => panic!(),
     }
 }
+
+/// Tests that one node forwards chunk parts correctly to other nodes.
+/// When a validator V receives a PartialEncodedChunk from the chunk producer,
+/// containing the parts that the V owns, it is responsible for forwarding
+/// these parts to the next chunk producer as well as the any block producer
+/// that tracks the shard, as an optimization so that they don't have to
+/// request what we already know they will request. This test checks that
+/// behavior.
+///
+/// Note that before Phase 2 of sharding is launched, we actually forward to
+/// all block producers, not just those tracking the shard, because all
+/// block producers are required to track all shards.
+///
+/// See the comment on top of chain/chunks/src/lib.rs for more details.
+#[test]
+fn test_chunk_forward() {
+    let builder = ShardsManagerTestLoopBuilder::new();
+    // For this test we need to ensure that 1 part is not enough to decode
+    // the whole chunk. For that, we need at least 7 block producers, so that
+    // the total number of parts is 7, and the number of parts needed to
+    // decode the chunk is 2, since the formula is (num_block_producers - 1)/3.
```

Comment:
So in this test we test the following:
1. A block producer who receives a PartialEncodedChunk will forward that part to all other block producers, and also to the next chunk producer; this is tested by capturing network messages and asserting that they are all there and there are no extras.
2. A block producer who receives a PartialEncodedChunk, who _would normally immediately send requests for any other parts/receipts that are missing_, does _not_ do so in this case because it is expecting a forward. This is done by asserting that the request message is not sent with run_instant() but only sent after run_for(CHUNK_REQUEST_RETRY).

So if we used <7 validators, then when a block producer B receives any PartialEncodedChunk (that contains 1 part -- the part that B owns), it would already be able to decode the entire chunk, and it won't even want to send any requests at all, so (2) is not testable.

This test does not assert anything from the perspective of chunk-only producers, because the ShardsManager instance we set up is for the block producer.

Reply:
(Note that the reason why this test exists in the first place is that in a multi-instance test, (2) also does not trigger. That's because the block producer would have received forwards from other block producers, and not need to send requests at all. The retry path for such chunks is only needed for block producers if there is network fault or high latency. In this single-instance test, we don't give this block producer any incoming forwarding message, so it would decide to send a request proactively after the retry timeout.)